### DISCUSSION TITLE
blackpill401cc - Preset change

### DISCUSSION THREAD

**Discussion by Wapata** [shanteacontrols/OpenDeck#186] [General]
Hi there !
It's been a long time but i've did it, a simple wind controller by using the power of OpenDeck.

So, it's blacpill401cc board, with pressure sensor into Analog 0 / A1, (plus a swicth to put A1 to 3.3volts to bypass the sensor) and a Button 2 / B10.
The job is simple, Button 2 should navigate from Preset 1 to 2 and vice versa, Analog 0 should be CC2 on preset 1 and CC11 on preset 2.
...
The button work great except if I configure it to "preset change" / "Midi ID : 2" (or 1) then the board send crazy messages.

Here is the response to Button 2 = note (first push), then Button 2 = Preset (second push):
`<html>
<body>

Time | Event
-- | --
20:47:59 16 | Note On 7channel 1velocity 0Raw data[ 144 7 0 ]  Dec
20:47:59 13 | Note On 6channel 1velocity 0Raw data[ 144 6 0 ]  Dec
20:47:59 10 | Note On 5channel 1velocity 0Raw data[ 144 5 0 ]  Dec
20:47:59 2 | Note On 4channel 1velocity 0Raw data[ 144 4 0 ]  Dec
20:47:59 0 | Note On 3channel 1velocity 0Raw data[ 144 3 0 ]  Dec
20:47:58 998 | Note On 2channel 1velocity 0Raw data[ 144 2 0 ]  Dec
20:47:58 995 | Component infoButton #1Raw data[ 240 0 83 67 1 0 73 1 0 1 247 ]  Dec
20:47:58 990 | Note On 1channel 1velocity 0Raw data[ 144 1 0 ]  Dec
20:47:58 988 | Note On 0channel 1velocity 0Raw data[ 144 0 0 ]  Dec
20:47:58 704 | Note On 2channel 1velocity 0Raw data[ 144 2 0 ]  Dec
20:47:58 489 | Component infoButton #2Raw data[ 240 0 83 67 1 0 73 1 0 2 247 ]  Dec
20:47:58 488 | Note On 2channel 1velocity 127Raw data[ 144 2 127 ]  Dec
20:47:54 37 | 118.  SetValue- Button # 2 Message type17donein11mswaited4msSent[ 240 0 83 67 0 0 1 0 1 1 0 2 0 17 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[]
20:47:44 745 | Component infoButton #2Raw data[ 240 0 83 67 1 0 73 1 0 2 247 ]  Dec
20:47:44 744 | Note On 2channel 1velocity 0Raw data[ 144 2 0 ]  Dec
20:47:44 203 | Component infoButton #2Raw data[ 240 0 83 67 1 0 73 1 0 2 247 ]  Dec
20:47:44 198 | Note On 2channel 1velocity 127Raw data[ 144 2 127 ]  Dec

</body>
</html>`

![IMG_20221231_201525_322](https://user-images.githubusercontent.com/15179434/210154368-d1f20b46-ede3-4ff6-b15c-b43071254219.jpg)

![IMG_20221231_201519_083](https://user-images.githubusercontent.com/15179434/210154369-b5275395-a963-4d06-8542-5a3384230efb.jpg)

**Comment by paradajz** [comment]
Once the preset is changed, the board will resend all values for buttons and enabled potentiometers. Same thing happens once the USB connection is established.

**Reply by Wapata** [reply]
Okay then, so it's a matter of puting to "none" all the switches except n°2 for me.
... All the 15 switches for the two presets :-p

I'll try that and see if it's okay !

**Reply by Wapata** [reply]
Weel... it doesn't work.
Preset 1 button 2 is Note On, Preset 2 button 2 is "get back to preset one"
here is the result after touching button 2 into preset 1 and then preset 2
`<html>
<body>
<div class="app-main" style="box-sizing: border-box; border: 0px solid rgb(224, 224, 224); font-size: 1.125rem; width: 1586px; color: rgb(224, 224, 224); font-family: system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, &quot;Noto Sans&quot;, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="content" style="box-sizing: border-box; border: 0px solid rgb(224, 224, 224);"><div class="request-log section" style="box-sizing: border-box; border-width: 1px 0px 0px; border-style: solid; border-color: rgba(52,52,52,var(--border-opacity)); border-image: initial; width: 1586px; --bg-opacity:1; background-color: rgba(0,0,0,var(--bg-opacity)); --text-opacity:1; color: rgba(189,189,189,var(--text-opacity)); --border-opacity:1;"><div class="request-log-main" style="box-sizing: border-box; border: 0px solid rgb(224, 224, 224); font-size: 0.875rem; overflow-y: scroll; max-height: 500px;">

Time | Event
-- | --
12:56:26 119 | 183.  GetValue- Global - DIN to USB Thrudonein2mswaited96msSent[ 240 0 83 67 0 0 0 0 0 0 0 2 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 118 | 182.  GetValue- Global - DIN to DIN Thrudonein2mswaited95msSent[ 240 0 83 67 0 0 0 0 0 0 0 7 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 117 | 181.  GetValue- Global - USB to USB Thrudonein17mswaited78msSent[ 240 0 83 67 0 0 0 0 0 0 0 5 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 117 | 180.  GetValue- Global - USB to DIN Thrudonein3mswaited75msSent[ 240 0 83 67 0 0 0 0 0 0 0 4 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 116 | 179.  GetValue- Global - DIN MIDIdonein2mswaited74msSent[ 240 0 83 67 0 0 0 0 0 0 0 3 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 116 | 178.  GetValue- Global - Running statusdonein2mswaited71msSent[ 240 0 83 67 0 0 0 0 0 0 0 1 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 115 | 177.  GetValue- Global - Standard note offdonein2mswaited70msSent[ 240 0 83 67 0 0 0 0 0 0 0 0 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 115 | 176.  GetValue- Global - Global channeldonein3mswaited67msSent[ 240 0 83 67 0 0 0 0 0 0 0 14 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 1 ]
12:56:26 114 | 175.  GetValue- Global - Use global channeldonein3mswaited65msSent[ 240 0 83 67 0 0 0 0 0 0 0 13 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:26 114 | 174.  GetValue- Global - Active presetdonein11mswaited54msSent[ 240 0 83 67 0 0 0 0 0 2 0 0 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 1 ]
12:56:26 113 | 173.  GetValue- Global - Preserve preset settingdonein39mswaited15msSent[ 240 0 83 67 0 0 0 0 0 2 0 1 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:13 275 | 172.  GetValue- Button # 2 Valuedonein2mswaited51msSent[ 240 0 83 67 0 0 0 0 1 3 0 2 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 127 ]
12:56:13 274 | 171.  GetValue- Button # 2 MIDI IDdonein1mswaited50msSent[ 240 0 83 67 0 0 0 0 1 2 0 2 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 1 ]
12:56:13 274 | 170.  GetValue- Button # 2 MIDI channeldonein2mswaited47msSent[ 240 0 83 67 0 0 0 0 1 4 0 2 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 1 ]
12:56:13 274 | 169.  GetValue- Button # 2 Message typedonein10mswaited36msSent[ 240 0 83 67 0 0 0 0 1 1 0 2 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 17 ]
12:56:13 273 | 168.  GetValue- Button # 2 Typedonein29mswaited7msSent[ 240 0 83 67 0 0 0 0 1 0 0 2 0 0 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[ 0 ]
12:56:03 157 | Component infoAnalog #0Raw data[ 240 0 83 67 1 0 73 3 0 0 247 ]  Dec
12:56:03 156 | Control Change channel 1controller 2 value 51Raw data[ 176 2 51 ]  Dec
12:55:51 29 | Component infoAnalog #0Raw data[ 240 0 83 67 1 0 73 3 0 0 247 ]  Dec
12:55:51 28 | Control Change channel 1controller 2 value 51Raw data[ 176 2 51 ]  Dec
12:55:24 307 | Component infoAnalog #0Raw data[ 240 0 83 67 1 0 73 3 0 0 247 ]  Dec
12:55:24 304 | Control Change channel 1controller 2 value 51Raw data[ 176 2 51 ]  Dec
12:55:23 796 | 167.  SetValue- Global - Active preset1donein9mswaited6msSent[ 240 0 83 67 0 0 1 0 0 2 0 0 0 1 247 ]  DecReceived[ 240 0 83 67 247 ]  DecParsed[]
12:55:16 413 | Note On 2channel 1velocity 0Raw data[ 144 2 0 ]  Dec
12:55:16 293 | Component infoButton #2Raw data[ 240 0 83 67 1 0 73 1 0 2 247 ]  Dec
12:55:16 288 | Note On 2channel 1velocity 127Raw data[ 144 2 127 ]  Dec

</div></div></div></div><div class="app-footer" style="box-sizing: border-box; border-width: 0px; border-style: solid; border-color: rgba(52,52,52,var(--border-opacity)); border-image: initial; align-self: flex-end; padding-bottom: 2rem; width: 1586px; --bg-opacity:1; background-color: rgb(23, 23, 23); --text-opacity:1; color: rgba(189,189,189,var(--text-opacity)); --border-opacity:1; display: flex; font-family: system-ui, -apple-system, &quot;Segoe UI&quot;, Roboto, Ubuntu, Cantarell, &quot;Noto Sans&quot;, sans-serif, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, &quot;Helvetica Neue&quot;, Arial, &quot;Noto Sans&quot;, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="app-footer-wrap" style="box-sizing: border-box; border: 0px solid rgb(224, 224, 224); margin-left: auto; margin-right: auto; max-width: 72rem; display: flex;"><nav class="app-about" style="box-sizing: border-box; border: 0px solid rgb(224, 224, 224); width: 576px; padding: 1rem 2rem;"><h3 class="heading" style="box-sizing: border-box; border: 0px solid rgb(224, 224, 224); margin: 0px; font-size: inherit; font-weight: 700; padding-bottom: 1rem;">About</h3></nav></div></div>
</body>
</html>`


What seem strange to me is that the "Component infoButton #2" message doesn't exist into tu Preset 2

**Reply by paradajz** [reply]
> Weel... it doesn't work.

What doesn't work? Can you please be more specific on what are your expectations? I don't understand the issue here.

**Reply by Wapata** [reply]
Yes of course.
One button wired : button 2 into b10 pin.
In preset 1 it should launch preset two, and vice versa. But it doesn't.
The only other thing wired is Analog 0 into Pin A1 (Cc2 or cc11).
All the others parameters for buttons and analog ore set to "none".
So my guess is that there is a problem with "preset" and this board.
... I can try the same thing on an other kind of board if it help.

**Comment by paradajz** [comment]
Relevant discussion:
https://github.com/shanteacontrols/OpenDeck/discussions/44

**Comment by paradajz** [comment]
This took some time to figure out. Can you try the attached firmware?
[blackpill401cc.zip](https://github.com/shanteacontrols/OpenDeck/files/10384110/blackpill401cc.zip)

**Reply by Wapata** [reply]
Whatever the time, I'm always super impressed by your results. I will try as soon as I came back home.
Side note : so you have absolutely all the boards you've made compatible with OpenDeck into your desk, ready to be tested ? What you do feel like magic to me...

**Reply by paradajz** [reply]
> Side note : so you have absolutely all the boards you've made compatible with OpenDeck into your desk, ready to be tested ?

Not really, I only have a few, but the firmware / application part is the same everywhere, so a bug will manifest itself on every single board (unless it's something low-level but usually it isn't).

**Reply by Wapata** [reply]
Hi !
At home again, i've tried to update the firmware from the config.shanteacontrols page , pushed the blackpill401cc.hex and... have the exact same behavior.
I've made a reset to factory settings and it's also the same. Pushing button 2 from Preset 1 or Preset 2 make that :

`<html>
<body>


Time | Event
-- | --
17:37:11 583 | Note On 7channel 1velocity 0Raw data[ 144 7 0 ]  Dec
17:37:11 581 | Note On 6channel 1velocity 0Raw data[ 144 6 0 ]  Dec
17:37:11 573 | Note On 5channel 1velocity 0Raw data[ 144 5 0 ]  Dec
17:37:11 571 | Note On 4channel 1velocity 0Raw data[ 144 4 0 ]  Dec
17:37:11 569 | Note On 3channel 1velocity 0Raw data[ 144 3 0 ]  Dec
17:37:11 567 | Note On 1channel 1velocity 0Raw data[ 144 1 0 ]  Dec
17:37:11 560 | Component infoButton #0Raw data[ 240 0 83 67 1 0 73 1 0 0 247 ]  Dec
17:37:11 559 | Note On 0channel 1velocity 0Raw data[ 144 0 0 ]  Dec
17:37:08 642 | Note On 7channel 1velocity 0Raw data[ 144 7 0 ]  Dec
17:37:08 641 | Note On 6channel 1velocity 0Raw data[ 144 6 0 ]  Dec
17:37:08 640 | Note On 5channel 1velocity 0Raw data[ 144 5 0 ]  Dec
17:37:08 639 | Note On 4channel 1velocity 0Raw data[ 144 4 0 ]  Dec
17:37:08 639 | Note On 3channel 1velocity 0Raw data[ 144 3 0 ]  Dec
17:37:08 638 | Note On 1channel 1velocity 0Raw data[ 144 1 0 ]  Dec
17:37:08 636 | Component infoButton #0Raw data[ 240 0 83 67 1 0 73 1 0 0 247 ]  Dec
17:37:08 635 | Note On 0channel 1velocity 0Raw data[ 144 0 0 ]  Dec


</body>
</html>`

**Reply by paradajz** [reply]
Yes but I've already told you that is the expected behavior - once you change the preset all the data for buttons and enabled analog components will be forcefully sent. So again, I don't know what are you expecting here? The preset is changed.

**Reply by Wapata** [reply]
Well it is not. It as jumped from preset 1 to 3 (and i haven't configured that) and i've pushed button 2 and it's written nowhere. Plus all the buttons are set to "none"
Maybe i will make a little video if you think it can be helpfull

**Reply by paradajz** [reply]
Yeah video will be helpful.

**Comment by Wapata** [comment]
Hi !
Just a note here that with the new 7.5.4 firmware it work. When I press the button, the preset change and the last value of the analog pin is sent.
But for a reason, the button on the web configurator doesn't blink blue anymore, nore the message "button 2 pressed" (or something similar) is written in the "activity" view.

Anyway, I just have to add an LED and a case now ! Thank you !

**Reply by paradajz** [reply]
> But for a reason, the button on the web configurator doesn't blink blue anymore, nore the message "button 2 pressed" (or something similar) is written in the "activity" view.

I assume "button 2" is the one that you've configured to change the preset? If so, the system events such as those don't show up in the configurator so what you're (not) seeing is expected.

**Reply by Wapata** [reply]
So perfect then ! 
Yeah !!!
Thank you a lot :-)
